llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py

```diff
@@ -0,0 +1,457 @@
+
+        # Check if the index exists. Throws ValueError if it doesn't
+        try:
+            self._check_index_exists()
+        except Exception as e:
+            raise
+
+        self._bucket = self._cluster.bucket(self._bucket_name)
+        self._scope = self._bucket.scope(self._scope_name)
+        self._collection = self._scope.collection(self._collection_name)
+
+        super().__init__()
+
+    def add(self, nodes: List[BaseNode], **kwargs: Any) -> List[str]:
+        """
+        Add nodes to the collection and return their document IDs.
+
+        Args:
+            nodes (List[BaseNode]): List of nodes to add.
+            **kwargs (Any): Additional keyword arguments.
+                batch_size (int): Size of the batch for batch insert.
+
+        Returns:
+            List[str]: List of document IDs for the added nodes.
+        """
+        from couchbase.exceptions import DocumentExistsException
+
+        batch_size = kwargs.get("batch_size", self.DEFAULT_BATCH_SIZE)
+        documents_to_insert = []
+        doc_ids = []
+
+        for node in nodes:
+            metadata = node_to_metadata_dict(
+                node,
+                remove_text=True,
+                text_field=self._text_key,
+                flat_metadata=self.flat_metadata,
+            )
+            doc_id: str = node.node_id
+
+            doc = {
+                self._text_key: node.get_content(metadata_mode=MetadataMode.NONE),
+                self._embedding_key: node.embedding,
+                self._metadata_key: metadata,
+            }
+
+            documents_to_insert.append({doc_id: doc})
+
+        for i in range(0, len(documents_to_insert), batch_size):
+            batch = documents_to_insert[i : i + batch_size]
+            try:
+                # convert the list of dicts to a single dict for batch insert
+                insert_batch = {}
+                for doc in batch:
+                    insert_batch.update(doc)
+
+                logger.debug("Inserting batch of documents to Couchbase", insert_batch)
+
+                # upsert the batch of documents into the collection
+                result = self._collection.upsert_multi(insert_batch)
+
+                logger.debug(f"Insert result: {result.all_ok}")
+                if result.all_ok:
+                    doc_ids.extend(insert_batch.keys())
+
+            except DocumentExistsException as e:
+                logger.debug(f"Document already exists: {e}")
+
+            logger.debug("Inserted batch of documents to Couchbase")
+        return doc_ids
+
+    def delete(self, ref_doc_id: str, **kwargs: Any) -> None:
+        """
+        Delete a document by its reference document ID.
+
+        :param ref_doc_id: The reference document ID to be deleted.
+        :param **kwargs: Additional keyword arguments.
+        :type ref_doc_id: str
+        :return: None
+        :rtype: None
+        """
+        try:
+            self._collection.remove(ref_doc_id)
```

Comment:
this is actually inteded to delete any node that has `ref_doc_id == ref_doc_id` in its metadata (so a filtered delete) -- can we support that here?

Reply:
Thanks, I did not realize that. Will adapt it.